### PR TITLE
Remove hipcc guards

### PR DESCRIPTION
**Pull request by aurianer**
Remove the device code guards introduced by #4947
- Undefine parts of the code using actions (assert false in the corresponding device code portion)
- Expand some macros (components and actions) to empty for device code

TODO:

- [ ] ~~Remove the guards in the tests and examples~~ (in another PR)

### CODE REVIEW

**Review by hkaiser — COMMENTED** [review]
Could you explain please what has changed such that you were able to remove the guards, please?

File: src/runtime/agas/detail/hosted_locality_namespace.cpp [hunk 1]
```diff
@@ -43,19 +42,29 @@ namespace hpx { namespace agas { namespace detail
...
     }
 
     std::vector<std::uint32_t> hosted_locality_namespace::localities()
     {
+#if !defined(HPX_COMPUTE_DEVICE_CODE)
         server::locality_namespace::localities_action action;
         return action(gid_);
+#else
+        HPX_ASSERT(false);
+        return std::vector<std::uint32_t>{1, std::uint32_t(0)};
```
Comment (msimberg) [inline src/runtime/agas/detail/hosted_locality_namespace.cpp:60]: This can be an empty vector.

**Review by msimberg — CHANGES_REQUESTED** [review]
Thanks @aurianer. Looking good already, I'd only make a few small changes, but they are minor and not very important.

**Comment by msimberg** [comment]
> Could you explain please what has changed such that you were able to remove the guards, please?

@aurianer please expand as necessary. Nothing has changed really. We knew all along that it would be possible to remove them. We just didn't know how difficult it would be and didn't want to block the first iteration of HIP support on resolving that. Turns out that it was easier than we thought, or @aurianer made at least made it look easier ;)

**Comment by aurianer** [comment]
> Could you explain please what has changed such that you were able to remove the guards, please?

Yes I edited the description of the PR :)

File: components/containers/unordered/include/hpx/components/containers/unordered/partition_unordered_map_component.hpp [hunk 2]
```diff
@@ -536,9 +543,14 @@ namespace hpx
         ///
         future<T> get_value(Key const& pos, bool erase) const
         {
+#if !defined(HPX_COMPUTE_DEVICE_CODE)
             HPX_ASSERT(this->get_id());
             return hpx::async<typename server_type::get_value_action>(
                 this->get_id(), pos, erase);
+#else
+            HPX_ASSERT(false);
+            return ::hpx::make_ready_future(T{});
```
[the remark this answers is not in the document]
Comment (aurianer) [reply components/containers/unordered/include/hpx/components/containers/unordered/partition_unordered_map_component.hpp:552]: For the `::hpx` I think it is already used [sometimes](https://gist.github.com/aurianer/3bb281efcdd6499cd9954ef6ae7f943e). Does the compiler search first for symbols in the current namespace or in the no namespace first? I was wondering, as we use `hpx::<something>` symbols quite a lot, if it makes a little difference on the overall codebase, but it's probably nothing so I can totally put `hpx`
Should I also replace the ones that are already on master?

[the remark this answers is not in the document]
Comment (msimberg) [reply components/containers/unordered/include/hpx/components/containers/unordered/partition_unordered_map_component.hpp:552]: Hmm, indeed. I suspect we still use `hpx` more often than `::hpx`, so let's stick with it for now. The lookup with `hpx` starts in the current namespace and then successively moves to parent namespaces, which in this case ends up at the top level anyway, so using `::hpx` might be more appropriate eventually. Let's keep the changes here minimal though.

**Review by msimberg — APPROVED** [review]
Thanks @aurianer!

**Comment by aurianer** [comment]
retest lsu

**Pull request merged**